crates/bevy_render/src/renderer/mod.rs

```diff
@@ -57,23 +57,104 @@ pub fn render_system(world: &mut World) {
-        .expect("Unable to find a GPU! Make sure you have installed required drivers!");
+) -> (RenderDevice, RenderAdapter, RenderQueue) {
+    let adapter = {
+        #[cfg(not(target_arch = "wasm32"))]
+        {
+            use crate::texture::{BevyDefault as _, DEFAULT_DEPTH_FORMAT};
+            let adapters: Vec<wgpu::Adapter> = instance.enumerate_adapters(backends).collect();
+            let mut selected_adapter = None;
+            let (mut integrated, mut discrete, mut virt, mut cpu, mut other) =
+                (None, None, None, None, None);
+
+            for (i, adapter) in adapters.iter().enumerate() {
+                let default_texture_format_features =
+                    adapter.get_texture_format_features(wgpu::TextureFormat::bevy_default());
+                let default_depth_format_features =
+                    adapter.get_texture_format_features(DEFAULT_DEPTH_FORMAT);
 
-    info!("{:?}", adapter.get_info());
+                if default_texture_format_features
+                    .allowed_usages
+                    .contains(wgpu::TextureUsages::RENDER_ATTACHMENT)
+                    && default_depth_format_features
+                        .allowed_usages
+                        .contains(wgpu::TextureUsages::RENDER_ATTACHMENT)
+                {
+                    let info = adapter.get_info();
+                    match info.device_type {
+                        wgpu::DeviceType::IntegratedGpu => {
+                            integrated = integrated.or(Some(i));
+                        }
+                        wgpu::DeviceType::DiscreteGpu => {
+                            discrete = discrete.or(Some(i));
+                        }
+                        wgpu::DeviceType::VirtualGpu => {
+                            virt = virt.or(Some(i));
+                        }
+                        wgpu::DeviceType::Cpu => {
+                            cpu = cpu.or(Some(i));
+                        }
+                        wgpu::DeviceType::Other => {
+                            other = other.or(Some(i));
+                        }
+                    }
+                }
+            }
+
+            let preferred_gpu_index = match options.power_preference {
+                wgpu::PowerPreference::LowPower => {
+                    integrated.or(other).or(discrete).or(virt).or(cpu)
+                }
+                wgpu::PowerPreference::HighPerformance => {
+                    discrete.or(other).or(integrated).or(virt).or(cpu)
+                }
+            }
+            .expect("Unable to find a GPU! Make sure you have installed required drivers!");
+
+            for (i, adapter) in adapters.into_iter().enumerate() {
+                if preferred_gpu_index == i {
+                    selected_adapter = Some(adapter);
+                }
+            }
+
+            selected_adapter
+                .expect("Unable to find a GPU! Make sure you have installed required drivers!")
```

Comment:
```suggestion
            adapters[i]
```

Reply:
borrow checker fight :)

```
error[E0507]: cannot move out of index of `Vec<Adapter>`
   --> crates\bevy_render\src\renderer\mod.rs:127:12
    |
127 |            adapters[preferred_gpu_index]
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |            |
    |            move occurs because value has type `Adapter`, which does not implement the `Copy` trait
    |            help: consider borrowing here: `&adapters[preferred_gpu_index]`
```

Reply:
Fair enough, sorry. You need to use `swap_remove` instead.